arangod/Aql/WalkerWorker.h

```diff
@@ -51,15 +48,20 @@ class WalkerWorker {
   virtual void after(T*) {}
 
   /// @brief return true to enter subqueries, false otherwise
-  virtual bool enterSubquery(T*, T*) {  // super, sub
-    return true;
-  }
+  virtual bool enterSubquery(T* /*super*/, T* /*sub*/) { return true; }
 
-  virtual void leaveSubquery(T*,  // super,
-                             T*   // sub
-  ) {}
+  virtual void leaveSubquery(T* /*super*/, T* /*sub*/) {}
 
-  virtual bool done(T* en) {
+  virtual bool done(T* en) { return false; }
+};
+
+/// @brief functionality to walk an execution plan recursively.
+/// this will fail in maintainer mode with assertion failure if the same 
+/// node is visited twice.
+template <class T>
```

Comment:
Maybe that will work. But it is a bit more complicated: the behavior of the WalkerWorker also depends on the value of the  `ARANGODB_ENABLE_MAINTAINER_MODE` define.
So we actually have 3 behaviors:
* a WalkerWorker that asserts on duplicates
* a WalkerWorker that visits duplicates multiple times
* a WalkerWorker that visits duplicates only once
I am pretty sure it can be cleaned up further